nlprule/src/components/tokenizer/mod.rs

```diff
@@ -0,0 +1,238 @@
+    #[serde(default)]
+    pub extra_join_regexes: Vec<Regex>,
+}
+
+/// The complete Tokenizer doing tagging, chunking and disambiguation.
+#[derive(Serialize, Deserialize, Default, Clone)]
+pub struct Tokenizer {
+    whitelist: DefaultHashSet<String>,
+    sentencizer: srx::Rules,
+    tagger: Tagger,
+    lang_options: TokenizerLangOptions,
+    #[serde(skip)]
+    properties: OnceCell<PropertiesMut>,
+}
+
+impl Tokenize for Tokenizer {
+    fn properties(&self) -> PropertiesMut {
+        lazy_static! {
+            static ref PROPERTIES: PropertiesMut = Properties::default().write(&[Property::Tags]);
+        }
+        *PROPERTIES
+    }
+
+    fn tokenize<'t>(&'t self, text: &'t str) -> Box<dyn Iterator<Item = Sentence<'t>> + 't> {
+        Box::new(SentenceIter {
+            text,
+            splits: self.sentencizer.split_ranges(text),
+            tokenizer: &self,
+            index: 0,
+            position: Position::default(),
+        })
+    }
+
+    fn tokenize_sentence<'t>(&'t self, sentence: &'t str) -> Option<Sentence<'t>> {
+        if sentence.trim().is_empty() {
+            return None;
+        }
+
+        let token_strs = self
+            .get_token_ranges(sentence)
+            .filter(|range| !sentence[range.clone()].trim().is_empty());
+
+        let n_token_strs = token_strs.clone().count();
+
+        let tokens: Vec<_> = token_strs
+            .enumerate()
+            .map(|(i, range)| {
+                let byte_start = range.start;
+                let char_start = sentence[..byte_start].chars().count();
+
+                let token_text = sentence[range].trim();
+
+                let is_sentence_start = i == 0;
+                let is_sentence_end = i == n_token_strs - 1;
+
+                Token::new(
+                    token_text,
+                    Span::new(
+                        byte_start..byte_start + token_text.len(),
+                        char_start..char_start + token_text.chars().count(),
+                    ),
+                    is_sentence_start,
+                    is_sentence_end,
+                    sentence[..byte_start].ends_with(char::is_whitespace),
+                )
+            })
+            .collect();
+
+        let mut sentence = Sentence::new(tokens, sentence, &self.tagger);
+        let guard = self.property_guard(&mut sentence).expect("TODO");
+
+        sentence = self.tagger.transform(sentence, guard).expect("TODO");
+
+        Some(sentence)
+    }
+}
+
+/// An iterator over sentences. Has some key properties:
+/// - Preceding whitespace is always included so the first sentence always starts at byte and char index zero.
+/// - There are no gaps between sentences i.e. `sentence[i - 1].span().end() == sentence[i].span().start()`.
+/// - Behavior for trailing whitespace is not defined. Can be included in the last sentence or not be part of any sentence.
+pub struct SentenceIter<'t> {
+    text: &'t str,
+    splits: Vec<Range<usize>>,
```

Comment:
I'd recommend to use two type aliases: `CharRange` and `ByteRange` to disambiguate throughout the code.